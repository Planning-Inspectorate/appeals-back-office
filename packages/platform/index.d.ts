import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
