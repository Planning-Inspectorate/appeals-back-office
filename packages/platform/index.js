export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
