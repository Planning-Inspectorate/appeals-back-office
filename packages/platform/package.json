{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
