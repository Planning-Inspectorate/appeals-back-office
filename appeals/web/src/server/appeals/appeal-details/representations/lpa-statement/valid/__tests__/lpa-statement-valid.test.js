import {
	appealDataFullPlanning,
	lpaStatementAwaitingReview,
	getAppealRepsResponse,
	allocationDetailsData
} from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';
import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';

const { app, installMockApi, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details';
const appealId = 2;

describe('lpa statement valid', () => {
	beforeEach(() => {
		installMockApi();

		nock('http://test/')
			.get(`/appeals/${appealId}/reps?type=lpa_statement`)
			.reply(200, {
				...getAppealRepsResponse,
				itemCount: 1,
				items: [
					{
						...lpaStatementAwaitingReview
					}
				]
			})
			.persist();
		nock('http://test/')
			.get('/appeals/appeal-allocation-levels')
			.reply(200, allocationDetailsData.levels)
			.persist();
		nock('http://test/')
			.get('/appeals/appeal-allocation-specialisms')
			.reply(200, allocationDetailsData.specialisms)
			.persist();
	});

	afterEach(() => {
		teardown;
		nock.cleanAll();
	});

	describe('GET /allocation-check', () => {
		it('should render the allocation check page with expected content', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					allocationDetails: null,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-check`
			);
			expect(response.statusCode).toBe(200);
			const { innerHTML } = parseHtml(response.text);
			expect(innerHTML).toMatchSnapshot();

			expect(innerHTML).toContain('Appeal 351062</span>');
			expect(innerHTML).toContain('Allocation level</h1>');
			expect(innerHTML).toContain(
				'Do you need to update the allocation level and specialisms?</legend>'
			);
			expect(innerHTML).toContain('name="allocationLevelAndSpecialisms" type="radio" value="yes">');
			expect(innerHTML).toContain('name="allocationLevelAndSpecialisms" type="radio" value="no">');
			expect(innerHTML).toContain('Continue</button>');
		});

		it('should render a back link to the review LPA statement page', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-check`
			);

			expect(response.statusCode).toBe(200);

			const unprettifiedHtml = parseHtml(response.text, {
				rootElement: 'body',
				skipPrettyPrint: true
			}).innerHTML;

			expect(unprettifiedHtml).toContain(
				`href="/appeals-service/appeal-details/${appealId}/lpa-statement" class="govuk-back-link">Back</a>`
			);
		});
	});

	describe('GET /allocation-level', () => {
		it('should render the allocation level page with expected content', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					allocationDetails: null,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-level`
			);
			expect(response.statusCode).toBe(200);
			const { innerHTML } = parseHtml(response.text);
			expect(innerHTML).toMatchSnapshot();

			expect(innerHTML).toContain('Appeal 351062</span>');
			expect(innerHTML).toContain(`Allocation level</h1>`);
			expect(innerHTML).toContain(
				`<label class="govuk-label govuk-radios__label" for="allocationLevel">A</label>`
			);
			expect(innerHTML).toContain(
				`<label class="govuk-label govuk-radios__label" for="allocationLevel-2">B</label>`
			);
			expect(innerHTML).toContain('Continue</button>');
		});

		it('should render a back link to the review LPA statement page, if the appeal has no allocation level or allocation specialisms (i.e. review page was previous page, allocation check was skipped because user must update allocation info)', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					allocationDetails: null,
					appealId,
					appealStatus: 'statements'
				});

			const reviewPageResponse = await request.post(`${baseUrl}/${appealId}/lpa-statement`).send({
				status: 'valid'
			});

			expect(reviewPageResponse.statusCode).toBe(302);

			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					allocationDetails: null,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-level`
			);

			expect(response.statusCode).toBe(200);

			const unprettifiedHtml = parseHtml(response.text, {
				rootElement: 'body',
				skipPrettyPrint: true
			}).innerHTML;

			expect(unprettifiedHtml).toContain(
				`href="/appeals-service/appeal-details/${appealId}/lpa-statement" class="govuk-back-link">Back</a>`
			);
		});

		it('should render a back link to the allocation check page, if the appeal has both an allocation level and allocation specialisms (i.e. allocation check was previous page because user was given the option to update allocation info or not)', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					appealId,
					appealStatus: 'statements'
				});

			const reviewPageResponse = await request.post(`${baseUrl}/${appealId}/lpa-statement`).send({
				status: 'valid'
			});

			expect(reviewPageResponse.statusCode).toBe(302);

			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-level`
			);

			expect(response.statusCode).toBe(200);

			const unprettifiedHtml = parseHtml(response.text, {
				rootElement: 'body',
				skipPrettyPrint: true
			}).innerHTML;

			expect(unprettifiedHtml).toContain(
				`href="/appeals-service/appeal-details/${appealId}/lpa-statement/valid/allocation-check" class="govuk-back-link">Back</a>`
			);
		});
	});

	describe('GET /allocation-specialisms', () => {
		it('should render the allocation specialisms page with expected content', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-specialisms`
			);
			expect(response.statusCode).toBe(200);

			const { innerHTML } = parseHtml(response.text);

			expect(innerHTML).toMatchSnapshot();
			expect(innerHTML).toContain('Appeal 351062</span>');
			expect(innerHTML).toContain(`Allocation specialisms</h1>`);
			expect(innerHTML).toContain(
				`<label class="govuk-label govuk-checkboxes__label" for="allocationSpecialisms">Specialism 1</label>`
			);
			expect(innerHTML).toContain(
				`<label class="govuk-label govuk-checkboxes__label" for="allocationSpecialisms-2">Specialism 2</label>`
			);
			expect(innerHTML).toContain(
				`<label class="govuk-label govuk-checkboxes__label" for="allocationSpecialisms-3">Specialism 3</label>`
			);
			expect(innerHTML).toContain('Continue</button>');
		});

		it('should render a back link to the valid journey version of the allocation level page', async () => {
			nock('http://test/')
				.get(`/appeals/${appealId}`)
				.reply(200, {
					...appealDataFullPlanning,
					appealId,
					appealStatus: 'statements'
				});

			const response = await request.get(
				`${baseUrl}/${appealId}/lpa-statement/valid/allocation-specialisms`
			);
			expect(response.statusCode).toBe(200);

			const { innerHTML } = parseHtml(response.text, {
				rootElement: 'body',
				skipPrettyPrint: true
			});

			expect(innerHTML).toContain(
				`href="/appeals-service/appeal-details/${appealId}/lpa-statement/valid/allocation-level" class="govuk-back-link"`
			);
		});
	});
});
