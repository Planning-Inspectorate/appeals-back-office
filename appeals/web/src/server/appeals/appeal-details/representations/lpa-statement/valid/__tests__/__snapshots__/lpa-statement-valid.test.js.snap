// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lpa statement valid GET /allocation-check should render the allocation check page with expected content 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Allocation level</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <fieldset class="govuk-fieldset">
                                <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">Do you need to update the allocation level and specialisms?</legend>
                                <div                                 class="govuk-radios" data-module="govuk-radios">
                                    <div class="govuk-radios__item">
                                        <input class="govuk-radios__input" id="allocationLevelAndSpecialisms"
                                        name="allocationLevelAndSpecialisms" type="radio" value="yes">
                                        <label class="govuk-label govuk-radios__label" for="allocationLevelAndSpecialisms">Yes</label>
                                    </div>
                                    <div class="govuk-radios__item">
                                        <input class="govuk-radios__input" id="allocationLevelAndSpecialisms-2"
                                        name="allocationLevelAndSpecialisms" type="radio" value="no">
                                        <label class="govuk-label govuk-radios__label" for="allocationLevelAndSpecialisms-2">No</label>
                                    </div>
                        </div>
                        </fieldset>
                </div>
                <button type="submit" data-prevent-double-click="true" class="govuk-button"
                data-module="govuk-button">Continue</button>
                </form>
            </div>
        </div>
    </div>
    </div>
</main>"
`;

exports[`lpa statement valid GET /allocation-level should render the allocation level page with expected content 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Allocation level</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <div class="govuk-radios" data-module="govuk-radios">
                                <div class="govuk-radios__item">
                                    <input class="govuk-radios__input" id="allocationLevel" name="allocationLevel"
                                    type="radio" value="A">
                                    <label class="govuk-label govuk-radios__label" for="allocationLevel">A</label>
                                </div>
                                <div class="govuk-radios__item">
                                    <input class="govuk-radios__input" id="allocationLevel-2" name="allocationLevel"
                                    type="radio" value="B">
                                    <label class="govuk-label govuk-radios__label" for="allocationLevel-2">B</label>
                                </div>
                            </div>
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;

exports[`lpa statement valid GET /allocation-specialisms should render the allocation specialisms page with expected content 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Allocation specialisms</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
            <div class="govuk-grid-row">
                <div class="govuk-grid-column-full">
                    <form method="POST" novalidate="novalidate">
                        <div class="govuk-form-group">
                            <div class="govuk-checkboxes" data-module="govuk-checkboxes">
                                <div class="govuk-checkboxes__item">
                                    <input class="govuk-checkboxes__input" id="allocationSpecialisms" name="allocationSpecialisms"
                                    type="checkbox" value="1">
                                    <label class="govuk-label govuk-checkboxes__label" for="allocationSpecialisms">Specialism 1</label>
                                </div>
                                <div class="govuk-checkboxes__item">
                                    <input class="govuk-checkboxes__input" id="allocationSpecialisms-2" name="allocationSpecialisms"
                                    type="checkbox" value="2">
                                    <label class="govuk-label govuk-checkboxes__label" for="allocationSpecialisms-2">Specialism 2</label>
                                </div>
                                <div class="govuk-checkboxes__item">
                                    <input class="govuk-checkboxes__input" id="allocationSpecialisms-3" name="allocationSpecialisms"
                                    type="checkbox" value="3">
                                    <label class="govuk-label govuk-checkboxes__label" for="allocationSpecialisms-3">Specialism 3</label>
                                </div>
                            </div>
                        </div>
                        <button type="submit" data-prevent-double-click="true" class="govuk-button"
                        data-module="govuk-button">Continue</button>
                    </form>
                </div>
            </div>
        </div>
    </div>
</main>"
`;
